feat / PLUG-44 - 리스트 컴포넌트 제작

GiToon10100011 opened the pull request:
**반영브랜치** 

- PLUG-44-ListComponent -> dev

**PR 타입**

- 기능 추가

**작업내용**

- 모드 기반 리스트 아이템 컴포넌트 구축
- 추천 리스트 및 재생목록 리스트 컴포넌트 구현
- 스타일 분리 및 재사용 패턴 적용

**사용예시**

```ts
// 1. 기본 리스트 아이템 사용
<ListItem 
  title="Meaning of you" 
  artist="아이유 IU" 
  mode="playlist" 
/>

// 2. 모드 전환을 통한 다양한 스타일 적용
<ListItem 
  title="어떻게 이별까지 사랑하겠어" 
  artist="AKMU (악뮤)" 
  mode="extract" 
/>

// 3. 추천 리스트 컴포넌트 사용
<RecoList 
  items={[
    { title: "너의 의미", artist: "아이유 IU", duration: "03:15" },
    { title: "주저하는 연인들을 위해", artist: "잔나비", duration: "04:26" }
  ]} 
/>

// 4. 유틸리티 함수를 통한 클래스명 관리
import { listItemModeSwitcher } from 'utils/modeSwitcher';

const className = listItemModeSwitcher('extract'); // "extract_item_container" 반환
```

mode는 `edit, extract, playlist, songs` 4개로 분류되어 있습니다.

**스크린샷**

- List 컴포넌트

<img width="1398" alt="image" src="https://github.com/user-attachments/assets/a8e20706-cb61-41f2-af3a-c5141859dea3" />


- RecoList 컴포넌트

<img width="1412" alt="image" src="https://github.com/user-attachments/assets/f1e8a40c-428a-4c4f-a071-367739396964" />

GiToon10100011 committed:
feat: 리스트 컴포넌트 시스템 구현 및 최적화

- ListItem 컴포넌트에 useCallback 메모이제이션 적용
- mode prop 기반 스타일 분기 처리 시스템 구축
- utils/modeSwitcher.ts 개발: 모드별 클래스명 중앙 관리
- utils/formatDuration.ts 구현: 재생 시간 표시 포맷팅 유틸리티
- 추천 리스트(RecoList) 컴포넌트 구현 및 스타일링
- 재생목록 리스트 종류 컴포넌트 클래스 분리로 스타일 확장성 확보

GiToon10100011 committed:
feat: 리스트 컴포넌트 테스트 페이지 제작

GiToon10100011 committed:
Merge branch 'dev' into PLUG-44-ListComponent

poan1221 commented on app/component/list/ListItem.tsx:
```diff
@@ -0,0 +1,48 @@
+          </p>
+        </div>
+        <div className={styles.container_rightside}>
+          <button
```
버튼은 희건님이 작성해주신 버튼 컴포넌트로 변경 부탁드립니다.
- dev 소스 머지해서 컴포넌트 사용하기
- 아마 icon 이미지들도 별도로 안넣어도 될 것 같습니다.

GiToon10100011 replied:
dev브랜치 현재 작업중인 브랜치에 그럼 머지해도 되는걸까요? 아니면 이것도 새로 브랜치 따서 작업을 해야하는걸까요?

poan1221 replied:
넵 지금 버전으로 머지해서 변경해주세요!

poan1221 commented on app/component/list/RecoListItem.module.scss:
```diff
@@ -0,0 +1,68 @@
+    padding-bottom: 20px;
+    display: flex;
+    justify-content: space-between;
+    align-items: center;
```
@mixin flex 로 구현되어 있는 부분 같아요!

poan1221 commented on app/component/list/RecoListItem.tsx:
```diff
@@ -0,0 +1,39 @@
+            height={50}
+          />
+          <p className={styles.desc_container}>
+            <span className={styles.title}>{title ?? "Meaning of you"}</span>
```
`title ?? "Meaning of you"` 로 쓰면, null, undefined 일 때는 기본 값이 들어가는데, '' 빈문자열일 때는 빈 걸로 들어갈 텐데 이 의도가 맞을까요?

빈문자열일 경우는 없으니까 괜찮을까요? 🤔
빈문자열까지 처리 하려면 || 연산 써야 할 것 같습니다.

GiToon10100011 replied:
빈문자열일 경우가 없을거라 생각해서 그렇게 만든거긴 한데 만약을 대비해 ||로 수정하는 것이 좋을까요?

poan1221 replied:
저도 없을 것 같긴 한데, 고민되는 포인트네용.. 없을거 같으니 일단 그냥 둘까요? 어디서 데이터를 가져오든 아마 title 은 있을 테니까요..?

GiToon10100011 replied:
넵 그냥 그냥 두는 방향으로 적용하겠습니다!

poan1221 commented on app/component/list/page.tsx:
```diff
@@ -0,0 +1,12 @@
+export default function ListCompTestPage() {
+  return (
+    <div>
+      <RecoList />
```
컴포넌트로 사용할 때, 반환된 데이터를 넣어줘야 하는데 인자는 어떻게 전달하는 걸까요?
- ui 컴포넌트이긴 한데, 데이터 넣는 곳이 아무곳도 없으면.. 이걸 어떻게 사용하지는 지에 대한 예시가 한번 더 있어야 할 것 같아요 (헷갈)

GiToon10100011 replied:
인자는 RecoList 컴포넌트 내부에서 데이터 페칭을 할거라 생각해서 따로 리스트에 대한 props를 정의하진 않았습니다!

poan1221 replied:
네! 그럼 해당 페이지에 있는 RecoList 랑 List 컴포넌트를 예시로 삼아서 실제 사용하는 페이지 내 컴포넌트로 만들어서 사용하는 걸로 하시죠.

poan1221 reviewed (commented):
확인 후 수정 부탁드립니다!
- 스타일 mixin 있는 걸로 사용
- 버튼 컴포넌트로 변경
- 빈문자열에 대한 고민 > 빈 것 그대로 반영할 것인지
- 컴포넌트로 사용 시 데이터 전달(?)

GiToon10100011 committed:
merge: dev 브랜치와 병합

GiToon10100011 committed:
style: 리스트 호버 스타일 수정

- flex 스타일 mixin 활용하여 수정
- button 요소 버튼 컴포넌트로 대체 및 불필요한 icon mixin 제거
- 그리드 레이아웃 전용 border-top 지역 믹스인 추가

GiToon10100011 committed:
Merge branch 'PLUG-44-ListComponent' of https://github.com/FRONT-END-BOOTCAMP-PULS-4/Jackpot_Plug into PLUG-44-ListComponent

GiToon10100011 committed:
style: hover 스타일 수정

- 호버 시 컨텐츠 보다 배경이 더 넘지 않도록 수정

GiToon10100011 committed:
merge: dev 브랜치와 병합

GiToon10100011 committed:
feat: 비디오 리스트 컴포넌트 추가

- utils/modeSwitcher.ts에 비디오 리스트 mode전환 추가
- 비디오 리스트(VideoList) 컴포넌트 구현 및 추가
- 재생목록 리스트(PlaylistList) 컴포넌트 구현 및 추추
- Button 컴포넌트에 xxs 옵션 추가
- 플레이리스트 삭제, 인증, 선택, 재생목록 추가 아이콘 추가

GiToon10100011 committed:
refactor: 함수 개선 및 리펙토링

- VideoList.tsx의 인덱스 상태관리 useCallback을 통해 최적화
- VideoListItem.tsx에서 조건부 렌더링을 renderMedia 함수로 리펙토링 하여 간소화
